web/universal_links.go

```diff
@@ -86,3 +86,41 @@ func universalLinkRedirect(c echo.Context) error {
 	}
 	return echo.NewHTTPError(http.StatusBadRequest, "This domain is not allowed to be redirected")
 }
+
+func webAuthRedirect(c echo.Context) error {
+	space, err := getSpaceFromHost(c)
+	if err != nil {
+		return err
+	}
+	spacePrefix := space.GetPrefix()
+	fallback := c.QueryParam("fallback")
+
+	if fallback == "" {
+		return echo.NewHTTPError(http.StatusNotFound)
+	}
+
+	// Disallow redirection for untrusted domains
+	parsedRedirect, err := url.Parse(fallback)
+	if err != nil {
+		return err
+	}
+
+	query := parsedRedirect.Query()
+	for k, v := range c.QueryParams() {
+		if k == "fallback" {
+			continue
+		}
+		query.Set(k, v[0])
+	}
+	parsedRedirect.RawQuery = query.Encode()
+
+	spaceTrustedDomains := base.Config.TrustedDomains
+	if domains, ok := spaceTrustedDomains[spacePrefix.String()]; ok {
+		for _, domain := range domains {
+			if strings.Contains(parsedRedirect.Host, domain) {
```

Comment:
Why is it `Contains` and not `HasSuffix`?

Reply:
I just copy the code from universal link. Perhaps we already have the trouble here :thinking:

Reply:
Yup, it looks like that